images/virtualization-artifact/pkg/controller/vd/internal/condition/ready.go

```diff
@@ -0,0 +1,93 @@
+package condition
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/deckhouse/virtualization-controller/pkg/common/vmd"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/supplements"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vdcondition"
+)
+
+type ReadyHandler struct {
+	diskService Disk
+}
+
+func NewReadyHandler(diskService Disk) *ReadyHandler {
+	return &ReadyHandler{
+		diskService: diskService,
+	}
+}
+
+func (h ReadyHandler) Handle(ctx context.Context, vd *virtv2.VirtualDisk) error {
+	condition, ok := getCondition(vdcondition.ReadyType, vd.Status.Conditions)
+	if !ok {
+		condition = metav1.Condition{
+			Type:   vdcondition.ReadyType,
+			Status: metav1.ConditionUnknown,
+		}
+	}
+
+	defer func() { setCondition(vd, condition) }()
+
+	if vd.DeletionTimestamp != nil {
+		condition.Status = metav1.ConditionUnknown
+		condition.Reason = ""
+		condition.Message = ""
+		return nil
+	}
+
+	supgen := supplements.NewGenerator(vmd.ShortName, vd.Name, vd.Namespace, vd.UID)
+
+	pvc, err := h.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return err
+	}
+
+	if condition.Status == metav1.ConditionTrue {
+		if pvc == nil {
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vdcondition.ReadyReason_Lost
+			condition.Message = fmt.Sprintf("Underling PersistentVolumeClaim %s is lost.", supgen.PersistentVolumeClaim().String())
+		}
+
+		return nil
+	}
+
+	if vd.Spec.DataSource != nil && vd.Spec.DataSource.Type != virtv2.DataSourceTypeObjectRef {
+		var provisioningPodStartedCondition metav1.Condition
+		provisioningPodStartedCondition, ok = getCondition(vdcondition.ProvisioningPodStartedType, vd.Status.Conditions)
+
+		if !ok || provisioningPodStartedCondition.Status != metav1.ConditionTrue {
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vdcondition.ReadyReason_ProvisioningPodNotStarted
+			condition.Message = "Provisioning Pod not started yet."
+			return nil
+		}
+
+		// TODO get errors.
+	}
+
+	if vd.Spec.DataSource != nil && vd.Spec.DataSource.Type == virtv2.DataSourceTypeUpload {
+		var userUploadStartedCondition metav1.Condition
+		userUploadStartedCondition, ok = getCondition(vdcondition.UserUploadStartedType, vd.Status.Conditions)
+
+		if !ok || userUploadStartedCondition.Status != metav1.ConditionTrue {
+			condition.Status = metav1.ConditionFalse
+			condition.Reason = vdcondition.ReadyReason_UserUploadNotStarted
+			condition.Message = "User upload not started yet."
+			return nil
+		}
+	}
```

Comment:
Oops, first condition is true for Upload datasource. Wrong condition may set for upload.